Use Solr OCR Highlighting Plugin in Search in Document Plugin

beatrycze-volk opened the pull request:
The main aim of this PR is to implement  Solr OCR Highlighting Plugin inside the existing Search in Document Plugin.

Tasks:

- [x] adjust SOLR schema
- [x] remap index fields or possible usage of indexes with different schema than default one
- [x] remap input fields names to make query parameters configurable
- [x] add backend classes for handling results returned from SOLR index
- [x] update SearchInDocumentTool.php and file and its template to pass all needed parameters
- [x] rewrite SearchInDocument,js file to handle search and processing of the result
   - [x] display result snippets
   - [x] highlight searched word in the full text
   - [x] highlight searched word in the image
- [x] update documentation for TypoScript config

Changes which should be discussed:
- [x] Access-Control-Allow-Origin
Currently in PageViewProxy it is set. It causes problem when this header is also configured in .htaccess.
- [x] replace of submit button with standard function button in search form
For unknown reason submit form button with prevent page reload was not working for the partner, so it got changes to standard button which calls a method. It work for the partner now, but it has some implications for e.g. pressing Enter after writing search term simply submits form without executing search function. One of the possible solutions would be also map the key to the function, but it is not sure that there are no other implications not detected currently.

beatrycze-volk committed:
Introduce solr highlighting

beatrycze-volk committed:
Use config for index remapped in SearchInDocumentTool

beatrycze-volk committed:
Use highlighting feature

beatrycze-volk committed:
Update js file to use new OCR highlighting

beatrycze-volk committed:
Change submit in search to click

beatrycze-volk committed:
Move search to own method

beatrycze-volk committed:
Update methods in JS of search in document plugin

beatrycze-volk committed:
Change configuration option fromisIndexRemapped to searchURL

Actually isIndexRemapped is not needed as fields are using config variables which have defined default values equal to fields names in default SOLR schema of Kitodo Presentation.

beatrycze-volk committed:
Make search input names configurable

beatrycze-volk committed:
Use config for documentIdUrlSchema

It is needed for cases when document id is passed by url instead by simple id used in index or database.

beatrycze-volk committed:
Update JS ad PHP files to use configured input names and document id URL schema, implement correctly
displaying of the links

albig commented:
What do I have to do to install/activate the OCR Highlighting Plugin? Does the code work, if it's not installed?

beatrycze-volk commented:
> What do I have to do to install/activate the OCR Highlighting Plugin? Does the code work, if it's not installed?

https://dbmdz.github.io/solr-ocrhighlighting/installation/

I would assume that it doesn't work if not installed. I have only tested with SOLR index which had the plugin installed.

beatrycze-volk committed:
Fix errors pointed by Codacy

albig commented:
> > What do I have to do to install/activate the OCR Highlighting Plugin? Does the code work, if it's not installed?
> 
> https://dbmdz.github.io/solr-ocrhighlighting/installation/
> 
> I would assume that it doesn't work if not installed. I have only tested with SOLR index which had the plugin installed.

Ok. Than we have to include this information to the documentation. 

And I will setup a testsystem with this Solr plugin. May take some days.

beatrycze-volk committed:
Add max amount of snippets and restore query->getHighlighting

Plugin has return only one snippet per page as default. Restore getHighlighting without assigning it to the variable.

beatrycze-volk committed:
Adjust URL generation for results' links

After this adjustments displaying and highlighting of the results in image works.

beatrycze-volk committed:
Add event listener for Enter pressed in the search field

beatrycze-volk commented:
Should I mark it as ready to merge or we also decide that some changes can't be merged in and I will create new branch in which I will implement new stuff without changing old ones (I mean way of submitting the form and cross origin access)?

beatrycze-volk committed:
Restore usage of submit button for submitting search

beatrycze-volk committed:
Restore CORS header in PageViewProxy

albig commented:
> https://dbmdz.github.io/solr-ocrhighlighting/installation/
> 
> I would assume that it doesn't work if not installed. I have only tested with SOLR index which had the plugin installed.

You are right. It doesn't work out-of-the box. The search plugin still finds no snippets of course.The searchInDocument fails with exception.

I still don't like the idea to make the solr-ocrhighlighting plugin obligatory. And I will investigate howto upgrade an existing installation.

albig commented on Classes/Common/SolrSearchResult/ResultDocument.php:
```diff
@@ -0,0 +1,259 @@
+        $this->page = $record[$fields['page']];
+
+        $highlightingForRecord = $highlighting[$record[$fields['id']]][$fields['fulltext']];
+        $this->snippetsForRecord = $highlightingForRecord['snippets'];
```
This maybe null and the getArrayByIndex() function will fail late.
```suggestion
        $this->snippetsForRecord = is_array($highlightingForRecord['snippets']) ? $highlightingForRecord['snippets'] : [];
```

beatrycze-volk replied:
Shouldn't this ternary operator have something between ?: or is it something specific to is_array and it doesn't need assignment?

albig replied:
I think, you're right. This would return "true" in case of a valid array. I've updated my suggestion.

albig commented on Classes/Common/SolrSearchResult/ResultDocument.php:
```diff
@@ -0,0 +1,259 @@
+        return $highlightsIds;
+    }
+
+    private function parseSnippets() {
```
Please add PHPDoc comments to private functions, too.

beatrycze-volk replied:
Added

albig commented on Configuration/ApacheSolr/configsets/dlf/conf/solrconfig.xml:
```diff
@@ -754,6 +756,21 @@
+          <str>highlight</str>
+      </arr>
+  </requestHandler>
+
```
The registration of the searchComponent ist missing in this solrconfig.xml. See https://dbmdz.github.io/solr-ocrhighlighting/installation/

Without this, it's not possible to create a core.

```
<!-- Add a new named search component that takes care of highlighting OCR field values. -->
  <searchComponent class="de.digitalcollections.solrocr.solr.OcrHighlightComponent"
                   name="ocrHighlight" />
```

albig reviewed (changes requested):
I didn't manage to get the Solr working with
* Solr 8.8
* solr-ocrhighlighting 0.5.0 (current release)
* schema.xml and solrconfig.xml from this PR
* updated extension configuration for solr (using default values)

On indexing a document, I end up with the following error

<pre>
org.apache.solr.common.SolrException: Exception writing document id 36LOG_0003 to the index; possible analysis error.
	at org.apache.solr.update.DirectUpdateHandler2.addDoc(DirectUpdateHandler2.java:256)
	at org.apache.solr.update.processor.RunUpdateProcessorFactory$RunUpdateProcessor.processAdd(RunUpdateProcessorFactory.java:73)
...
Caused by: java.lang.RuntimeException: Could not parse pointer: 
	at de.digitalcollections.solrocr.model.SourcePointer.parse(SourcePointer.java:81)
	at de.digitalcollections.solrocr.lucene.filters.ExternalUtf8ContentFilterFactory.create(ExternalUtf8ContentFilterFactory.java:42)
	at org.apache.solr.analysis.TokenizerChain.initReader(TokenizerChain.java:97)
...
</pre>

Please check again and tell me your system setup to reproduce.

beatrycze-volk committed:
Add missing search component to SOLR config

albig commented on Classes/Plugin/Tools/SearchInDocumentTool.php:
```diff
@@ -113,14 +105,68 @@ public function main($content, $conf)
+    {
+        $id = $this->doc->uid;
+
+        if (!empty($this->conf['documentIdUrlSchema'])) {
```
What's the purpose of this? Could you please provide an example string to verify the string splitting magic below?

beatrycze-volk replied:
Sometimes id of the document is actually not an id but a link to xml file. Then to search in the index the full link can't be used so it is needed to extract real id out of the full link.

Example:
https://dev-ddb.fiz-karlsruhe.de/api/items/F45KF24U7UIIVCUJJBV334PD7IPE7JBW/source/record -> https://dev-ddb.fiz-karlsruhe.de/api/items/*id*/source/record

sebastian-meyer replied:
I think we need to distinguish more cases here:
- The normal case would be documents which are properly indexed in the TYPO3 database. Those don't need any special treatment for their ID anyway.
- The "DFG-Viewer" case would be documents which are not indexed in the database **nor in the Solr index**. Normally those can be identified by the fact that they have an URL as ID.
- Then there is the special "Zeitungsportal" case where we have documents, which are not indexed in the database, but are indexed in the Solr index. Those have an URL as ID as well, but need special treatment to extract the indexed ID from the URL.

I suggest using the TYPO3 hook mechanism to handle all those cases:
- If a document has a numerical ID, no further treatment is necessary.
- If a document has a string ID, this string is processed through a hook.
  - If the hook returns an integer, the document is treated as indexed.
  - If the hook returns the string unchanged, the document is considered unindexed and ignored when searching. (This would be the default if no hook is registered, e. g. in "DFG-Viewer mode".)

The advantages would be:
- We don't need to add Zeitungportal-specific logic to the main extension. Instead the hook can be part of the supplemental Zeitungsportal extension.
- The solution is extensible, because you can register more than one hook to handle different kinds of identifiers. So anybody with a setup of a pre-existing Solr index could map their document IDs to the index IDs and re-use their Solr index. (This could even be interesting for SLUB, when we think about consolidating our different indices - CMR, Find, Kitodo, ...)
- Processing the ID string wouldn't be limited to splitting a string, because a hook could include more complex logic like regular expressions or even database lookups.

beatrycze-volk replied:
> I suggest using the TYPO3 hook mechanism to handle all those cases:

Do you mean https://typo3.org/article/how-to-use-existing-hooks-in-your-own-extension ?

> If a document has a string ID, this string is processed through a hook.
> If the hook returns an integer, the document is treated as indexed.
> If the hook returns the string unchanged, the document is considered unindexed and ignored when searching. (This would be the default if no hook is registered, e. g. in "DFG-Viewer mode".)

But how to convert string ID to an integer?

sebastian-meyer replied:
We actually already have some hooks in Kitodo.Presentation. A good example is the way how we extract the record ID from METS:
Normally we would take `/mets:mets[@OBJID]` as the document's unique record identifier: https://github.com/kitodo/kitodo-presentation/blob/f12eede1596112dc359678768e9a9463f0791788/Classes/Common/MetsDocument.php#L147-L150
But if this attribute is not set (which is the case for old versions of Kitodo.Production), we use hooks to have a fallback. The hook is registered in `ext_localconf.php`: https://github.com/kitodo/kitodo-presentation/blob/f12eede1596112dc359678768e9a9463f0791788/ext_localconf.php#L219 and called in `Classes/Common/MetsDocument.php`: https://github.com/kitodo/kitodo-presentation/blob/f12eede1596112dc359678768e9a9463f0791788/Classes/Common/MetsDocument.php#L151-L158
The hook itself can be found in `Classes/Hooks/KitodoProductionHacks.php`: https://github.com/kitodo/kitodo-presentation/blob/f12eede1596112dc359678768e9a9463f0791788/Classes/Hooks/KitodoProductionHacks.php#L37-L68

albig replied:
Hook is a possibility. I think the modern way would be to make some "middleware". But I haven't done this before. Hook support itself seem to be not deprecated, although lots of core hooks are.

The DFG-Viewer case is not relevant in the SearchInDocument tool. This tool won't work with it anyway. The DFG-Viewer has it's own SRU search plugin, which requires a SRU server of course.

I suggest to first test and documentate this PR including indexing (doesn't work yet) and searching for Kitodo.Presentation alone. If this works, we can discuss further at this point.

albig replied:
Please add at least a sample of the documentIdUrlSchema setting. Or is this part of the documentation?

albig replied:
This should be already part of #630. But I still need some (inline) documentation about what is going on here, why we are doing this and what string you expect in documentIdUrlSchema.

beatrycze-volk committed:
Check if passed id is number or numeric string

Previous check was converting numeric string to int if it was starting with number

beatrycze-volk commented:
@albig: Is your error similar to something like that:

`org.apache.solr.common.SolrException: Exception writing document id ocrdoc-1 to the index; possible analysis error: Illegal char <:> at index 5: https://digital.slub-dresden.de/data/kitodo/TheDarea_416971482-19100223/TheDarea_416971482-19100223_ocr/00000001.xml
	at org.apache.solr.update.DirectUpdateHandler2.addDoc(DirectUpdateHandler2.java:247)
	at org.apache.solr.update.processor.RunUpdateProcessor.processAdd(RunUpdateProcessorFactory.java:76)`

?

After research I have found out that indexing when XML files are stored in data folder of the core works correctly but when they are passed from external URL the error appears.

I have checked in the GItHub page of the plugin and found https://github.com/dbmdz/solr-ocrhighlighting/issues/49:

> Currently we only support loading field values/OCR fragments from the file system.

beatrycze-volk commented:
I'm sorry for this few rebase pushes, I have made a bit of mess and needed to fix it. Now all commits should be correct.

albig commented:
> @albig: Is your error similar to something like that:
> 
> `org.apache.solr.common.SolrException: Exception writing document id ocrdoc-1 to the index; possible analysis error: Illegal char <:> at index 5: https://digital.slub-dresden.de/data/kitodo/TheDarea_416971482-19100223/TheDarea_416971482-19100223_ocr/00000001.xml at org.apache.solr.update.DirectUpdateHandler2.addDoc(DirectUpdateHandler2.java:247) at org.apache.solr.update.processor.RunUpdateProcessor.processAdd(RunUpdateProcessorFactory.java:76)`
> 
> ?
> 
> After research I have found out that indexing when XML files are stored in data folder of the core works correctly but when they are passed from external URL the error appears.
> 
> I have checked in the GItHub page of the plugin and found [dbmdz/solr-ocrhighlighting#49](https://github.com/dbmdz/solr-ocrhighlighting/issues/49):
> 
> > Currently we only support loading field values/OCR fragments from the file system.

Yes, I think this was the error I've seen. So we have to wait for the fix in the ocrhighlighting plugin?

beatrycze-volk commented:
> Yes, I think this was the error I've seen. So we have to wait for the fix in the ocrhighlighting plugin?

Yes, of course if they are going to fix it actually. That's not so sure after reading the conversation in this issue.

Other possiblity is to store content of the xml files instead pointers to those files.

beatrycze-volk committed:
Remove indexing of full text from file

Source pointer can't be on any other server han SOLR, so it is not usable here.

beatrycze-volk committed:
Save to index full OCR

+ add working logging for document and indexing

beatrycze-volk committed:
Remove getRawText methods

beatrycze-volk committed:
Return false from save method if documents were not indexed

Either because of incorrect solr core or because some errors while indexing

albig commented on Classes/Common/Document.php:
```diff
@@ -48,6 +49,14 @@
+    /**
+     * This holds the logger
+     *
+     * @var LogManager
```
Good idea, but would you draw the changes to LogManager to a separate pull request? This is not related to the ocr highlighting.

albig commented on Classes/Common/Document.php:
```diff
@@ -667,54 +676,45 @@ protected function getRawTextFromXml($id)
+        return $fullText;
+    }
+
+    private function getTextFormat($fileContent)
```
Don't forget the method documentation.

albig replied:
Ping.

albig replied:
Ping again.

beatrycze-volk replied:
I'm not sure what is wrong with displaying of this part of the code in the Files changed view. There comment is missing, but in fact it is already added.

albig reviewed (changes requested):
On my setup, the current state of the pull request does index all pages. Inside the fulltext field, the complete ALTO XML is saved. Not sure, that this is what we want.

But anyway:

* there is no toplevel document indexed
* that's why the collection plugin doesn't work
* that's why the search plugin doesn't work
* that's why the searchInDocument plugin doesn't work

So, maybe my setup is broken and/or I am missing something. We should look together on this work.

![Solr Admin](https://user-images.githubusercontent.com/3354500/117323020-ce05ae80-ae8e-11eb-8ef0-5dfa03125233.png)

beatrycze-volk commented:
>  Inside the fulltext field, the complete ALTO XML is saved. Not sure, that this is what we want.
What do you mean? OCR is determined out of this content.

> there is no toplevel document indexed

That's not surprise, I have commented out indexing of documents without OCR as they were causing error mentioned in #502.

> that's why the searchInDocument plugin doesn't work

On my local environment with 3 indexed documents it works. By searching and finding snippets I have tested if part with indexing documents containing OCR works.

beatrycze-volk committed:
Fix after rebase class for logger in Document instance

beatrycze-volk committed:
Highlight more than one word in search

beatrycze-volk committed:
Preserve search phrase and get hit list after hit link was clicked

beatrycze-volk committed:
Index documents with empty fulltext field

beatrycze-volk committed:
Use coordinates inside the function for searching full text features

beatrycze-volk committed:
Fix Codacy errors

beatrycze-volk commented:
The last two Codacy errors can't be fixed: _Identifier 'tx_dlf_viewer' is not in camel case._

albig commented on Classes/Common/Document.php:
```diff
@@ -650,81 +648,72 @@ public function getPhysicalPage($logicalPage)
         $this->_getPhysicalStructure();
         // ... and extension configuration.
-        $extConf = GeneralUtility::makeInstance(ExtensionConfiguration::class)->get(self::$extKey);
+        $extConf = unserialize($GLOBALS['TYPO3_CONF_VARS']['EXT']['extConf'][self::$extKey]);
```
Why do you revert to the deprecated syntax?

beatrycze-volk replied:
My changes were written before new syntax was introduced. I will use new syntax :)

albig commented on Documentation/Plugins/Index.rst:
```diff
@@ -1087,7 +1087,11 @@ The fulltext is fetched and rendered by JavaScript into the `<div id="tx-dlf-ful
 
 
 Search in Document Tool
+<<<<<<< HEAD
```
Please solve the merge conflict.

beatrycze-volk replied:
Fixed

beatrycze-volk committed:
Fix request handler in solr config

albig commented on Resources/Public/Javascript/PageView/PageView.js:
```diff
@@ -349,11 +358,18 @@ dlfViewer.prototype.displayHighlightWord = function() {
+            value = urlParams[param];
+        }
+
+        var values = decodeURIComponent(value).split(';'),
```
```suggestion
        var values = decodeURIComponent(value).split(';');
```
This causes unfortunately an JavaScript exception which breaks the Pageview.

albig commented on Resources/Public/Javascript/PageView/SearchInDocument.js:
```diff
@@ -110,25 +318,22 @@ $(document).ready(function() {
         $('#tx-dlf-search-in-document-query').val('');
     });
+
+    triggerSearchAfterHitLoad();
```
This will be called on the listview, too and will end up in an JavaScript exception

```
Uncaught TypeError: Cannot read property 'split' of undefined
    at getBaseUrl (merged-0478f68d822d724a9d8e6d53edf60223.js?1625082852:4207)
    at triggerSearchAfterHitLoad (merged-0478f68d822d724a9d8e6d53edf60223.js?1625082852:4385)
    at HTMLDocument.<anonymous> (merged-0478f68d822d724a9d8e6d53edf60223.js?1625082852:4466)
```

beatrycze-volk replied:
Is SearchInDocument used also in list view?

albig replied:
Yes. All JavaScript ist loaded. That's what we changed recently: https://github.com/kitodo/kitodo-presentation/blob/master/Configuration/TypoScript/setup.txt

beatrycze-volk replied:
Ok, so I have added check for undefined before trying to split base url.

albig reviewed (changes requested):
It works better but is still a little surprising:

* documents with emtpy titles won't index. Most newspaper issues have no title field set.
* search-in-document is quite funny
   * opening the search form inserts some number (page numer?) and executes a search --> why?
   * the search itself seems to work
   * clicking on the search result doesn't change the page and has IMHO incorrect URLs like `https://ddev-kitodo-presentation.ddev.site/workview?=4&=4&=4&hl=14_23_1577_1754%3B14_23_688_1196&tx_dlf%5Bid%5D=3&tx_dlf%5Bpage%5D=3`
![Digitale Sammlungen  Allgemeine Zeitschrift für Textil-Industrie](https://user-images.githubusercontent.com/3354500/124033437-3a69dd80-d9fa-11eb-83c8-5cfeea4bfb47.png)

beatrycze-volk committed:
Use new syntax for getting of configuration

beatrycze-volk committed:
Fix documentations

beatrycze-volk committed:
Update Resources/Public/Javascript/PageView/PageView.js

Co-authored-by: Alexander Bigga <alexander.bigga@slub-dresden.de>

beatrycze-volk committed:
Check for undefined baseUrl and remove of passing hl param for highligting

beatrycze-volk commented:
> opening the search form inserts some number (page numer?) and executes a search --> why?

One of the requirements was that after moving from hit list of found documents to the certain document, it should execute this search again inside the document to get all found snippets in this document. The inserted value comes from URL parameter defined in configuration`plugin.tx_dlf_searchindocumenttool.queryInputName`. Default value for this configuration is  `tx_dlf[query]`. Could you check if you have something set up for this config?

> clicking on the search result doesn't change the page and has IMHO incorrect URLs like https://ddev-kitodo-presentation.ddev.site/workview?=4&=4&=4&hl=14_23_1577_1754%3B14_23_688_1196&tx_dlf%5Bid%5D=3&tx_dlf%5Bpage%5D=3

I would suspect that configuration`plugin.tx_dlf_searchindocumenttool.queryInputName` has empty value and it causes this weird inserted parameters.

I will add check for empty values to avoid such a situation.

beatrycze-volk committed:
Add checks for empty query parameters

Push them only if they are actually existing.

beatrycze-volk committed:
Fix rebase errors

Commit:
Merge branch 'master' into use-solr-highlighting

albig commented on Classes/Plugin/Tools/SearchInDocumentTool.php:
```diff
@@ -95,7 +95,7 @@ public function main($content, $conf)
         // Fill markers.
         $markerArray = [
-            '###ACTION_URL###' => $this->getActionUrl(),
+            '###ACTION_URL###' => getActionUrl(),
```
```suggestion
            '###ACTION_URL###' => $this->getActionUrl(),
```

albig commented on Documentation/Plugins/Index.rst:
```diff
@@ -1087,7 +1087,7 @@ The fulltext is fetched and rendered by JavaScript into the `<div id="tx-dlf-ful
 
 Search in Document Tool
-^^^^^^^^^^^^^^^^^^^^^^^
+
```
Why did you remove the heading format?

beatrycze-volk replied:
Restored, it looks that it was lost in merge conflict.

albig commented on Resources/Public/Javascript/PageView/SearchInDocument.js:
```diff
@@ -43,17 +43,233 @@ function resetStart() {
         // Send the data using post
         $.post(
-            "/",
+            "https://sdvtypo3ddbzeitungsportaldev.slub-dresden.de/",
```
No, not really ;-)

beatrycze-volk replied:
Yeah, it was specific for DDB, I will revert it ;)

albig reviewed (changes requested):
Please have a look at the comments. This is not yet working.

beatrycze-volk committed:
Update Classes/Plugin/Tools/SearchInDocumentTool.php

Co-authored-by: Alexander Bigga <alexander.bigga@slub-dresden.de>

beatrycze-volk committed:
Reverse not needed changes

beatrycze-volk committed:
Add comment for documentIdUrlSchema

albig reviewed (changes requested):
We're getting closer.

What works for me
* indexing documents with and without full text
* fulltext search in documents with searchInDocumentPlugin

Using the fulltext search of the search plugin fails on showing the text snippets. Instead of the found snippet, the listview plugin marker `###SUBPREVIEW###` is filled with `"HPOS="1492"/><String WC="0.743529439" CONTENT="<em>Textil</em>-Industrie." HEIGHT="27" WIDTH="302" VPOS="388""`. 

![fulltextkitodo](https://user-images.githubusercontent.com/3354500/127525056-f83c1438-eb5d-40a7-961c-b92925e4ec00.png)

This seems to be caused in DocumentList.php: https://github.com/beatrycze-volk/kitodo-presentation/blob/use-solr-highlighting/Classes/Common/DocumentList.php#L278

I think, the adaption to the searchInDocumentTool you made for the oce highlighting plugin is necessary at this point, too. To make the general fulltext search work again with text snippets.

Could you please have a look at it?

beatrycze-volk commented:
> We're get!ting closer.

Great!

> Using the fulltext search of the search plugin fails on showing the text snippets. Instead of the found snippet, the listview plugin marker `###SUBPREVIEW###` is filled with `"HPOS="1492"/><String WC="0.743529439" CONTENT="<em>Textil</em>-Industrie." HEIGHT="27" WIDTH="302" VPOS="388""`.
> 
> ![fulltextkitodo](https://user-images.githubusercontent.com/3354500/127525056-f83c1438-eb5d-40a7-961c-b92925e4ec00.png)
> 
> This seems to be caused in DocumentList.php: https://github.com/beatrycze-volk/kitodo-presentation/blob/use-solr-highlighting/Classes/Common/DocumentList.php#L278

Could you write down raw SOLR query which is created inside this method during search shown above?

albig commented:
> > This seems to be caused in DocumentList.php: https://github.com/beatrycze-volk/kitodo-presentation/blob/use-solr-highlighting/Classes/Common/DocumentList.php#L278
> 
> Could you write down raw SOLR query which is created inside this method during search shown above?

That's what Solr logs on my search for "Textil".

<pre>
2021-07-30 08:31:03.467 INFO  (qtp443721024-24) [   x:dlfCore0] o.a.s.c.S.Request [dlfCore0]  webapp=/solr path=/select params={q=fulltext:(textil)&json.nl=flat&omitHeader=true&fl=*,score&start=0&sort=score+desc&rows=0&wt=json} hits=294 status=0 QTime=0
2021-07-30 08:31:03.470 INFO  (qtp443721024-23) [   x:dlfCore0] o.a.s.c.S.Request [dlfCore0]  webapp=/solr path=/select params={q={!join+from%3Duid+to%3Duid}fulltext:(textil)&json.nl=flat&omitHeader=true&fl=uid,id&start=0&sort=score+desc&fq=toplevel:true&rows=25&wt=json} hits=6 status=0 QTime=1
2021-07-30 08:31:03.886 INFO  (qtp443721024-21) [   x:dlfCore0] o.a.s.c.S.Request [dlfCore0]  webapp=/solr path=/select params={json.nl=flat&hl=true&fl=uid,id,toplevel,thumbnail,page,type,title_tsi,volume_usu,author_tsi,place_tsi,year_usi&start=0&sort=score+desc&fq=uid:4&rows=10000&q=fulltext:(textil)+OR+toplevel:true&hl.useFastVectorHighlighter=true&omitHeader=true&hl.q=textil&hl.fl=fulltext&wt=json} hits=202 status=0 QTime=402
2021-07-30 08:31:04.305 INFO  (qtp443721024-22) [   x:dlfCore0] o.a.s.c.S.Request [dlfCore0]  webapp=/solr path=/select params={json.nl=flat&hl=true&fl=uid,id,toplevel,thumbnail,page,type,title_tsi,volume_usu,author_tsi,place_tsi,year_usi&start=0&sort=score+desc&fq=uid:3&rows=10000&q=fulltext:(textil)+OR+toplevel:true&hl.useFastVectorHighlighter=true&omitHeader=true&hl.q=textil&hl.fl=fulltext&wt=json} hits=25 status=0 QTime=50
2021-07-30 08:31:04.340 INFO  (qtp443721024-25) [   x:dlfCore0] o.a.s.c.S.Request [dlfCore0]  webapp=/solr path=/select params={json.nl=flat&hl=true&fl=uid,id,toplevel,thumbnail,page,type,title_tsi,volume_usu,author_tsi,place_tsi,year_usi&start=0&sort=score+desc&fq=uid:27&rows=10000&q=fulltext:(textil)+OR+toplevel:true&hl.useFastVectorHighlighter=true&omitHeader=true&hl.q=textil&hl.fl=fulltext&wt=json} hits=2 status=0 QTime=4
2021-07-30 08:31:04.438 INFO  (qtp443721024-24) [   x:dlfCore0] o.a.s.c.S.Request [dlfCore0]  webapp=/solr path=/select params={json.nl=flat&hl=true&fl=uid,id,toplevel,thumbnail,page,type,title_tsi,volume_usu,author_tsi,place_tsi,year_usi&start=0&sort=score+desc&fq=uid:5&rows=10000&q=fulltext:(textil)+OR+toplevel:true&hl.useFastVectorHighlighter=true&omitHeader=true&hl.q=textil&hl.fl=fulltext&wt=json} hits=57 status=0 QTime=92
2021-07-30 08:31:04.558 INFO  (qtp443721024-23) [   x:dlfCore0] o.a.s.c.S.Request [dlfCore0]  webapp=/solr path=/select params={json.nl=flat&hl=true&fl=uid,id,toplevel,thumbnail,page,type,title_tsi,volume_usu,author_tsi,place_tsi,year_usi&start=0&sort=score+desc&fq=uid:50&rows=10000&q=fulltext:(textil)+OR+toplevel:true&hl.useFastVectorHighlighter=true&omitHeader=true&hl.q=textil&hl.fl=fulltext&wt=json} hits=2 status=0 QTime=4
2021-07-30 08:31:04.584 INFO  (qtp443721024-21) [   x:dlfCore0] o.a.s.c.S.Request [dlfCore0]  webapp=/solr path=/select params={json.nl=flat&hl=true&fl=uid,id,toplevel,thumbnail,page,type,title_tsi,volume_usu,author_tsi,place_tsi,year_usi&start=0&sort=score+desc&fq=uid:60&rows=10000&q=fulltext:(textil)+OR+toplevel:true&hl.useFastVectorHighlighter=true&omitHeader=true&hl.q=textil&hl.fl=fulltext&wt=json} hits=12 status=0 QTime=20

</pre>

You can easily reproduce this with
* a page containing the search plugin
* a page containing the listview plugin

beatrycze-volk committed:
Use SOLR plugin in the document list

beatrycze-volk commented:
> That's what Solr logs on my search for "Textil".

Thanks!

> You can easily reproduce this with
> 
> * a page containing the search plugin
> * a page containing the listview plugin

Ok, I will need to get me page with this plugins on Monday.

 For now I have pushed the changes which should theoretically fix the problem, but it's not yet tested.

albig commented on Classes/Common/DocumentList.php:
```diff
@@ -316,6 +250,100 @@ protected function getRecord($element)
+            $query->addFilterQuery($this->metadata['options']['params']['filterquery']);
+        }
+        // Extend filter query to get all documents with the same UID.
+        foreach ($params['filterquery'] as $key => $value) {
```
That's true what Scrutinizer tells us. $params is uninitialized and the foreach ends in an exception.

albig commented on Classes/Common/DocumentList.php:
```diff
@@ -316,6 +250,100 @@ protected function getRecord($element)
         return $record;
     }
 
+    private function getSolrResult($record) {
```
Please add the PHPDocs.

albig reviewed (changes requested):
Unfortunately, your fast fix doesn't work yet. Please have a try after setting up your development environment.

Commit:
Merge branch 'master' into use-solr-highlighting

beatrycze-volk committed:
Fix errors and add missing docs

beatrycze-volk committed:
Merge remote-tracking branch 'origin/use-solr-highlighting' into use-solr-highlighting

beatrycze-volk committed:
Fix Scrutinizer warnings

albig commented on Classes/Common/DocumentList.php:
```diff
@@ -316,6 +252,122 @@ protected function getRecord($element)
+        if ($this->metadata['fulltextSearch']) {
+            $query->setQuery(
+                $fields['fulltext'] . ':' . Solr::escapeQuery($this->metadata['searchString']));
+                // TODO:toplevel documents throw error in SOLR PLugin:
```
I'm not sure, if this is the reason for the empty toplevel element in listview. Could you please have a look at this:

![Listview](https://user-images.githubusercontent.com/3354500/129206463-6962b640-2c4e-4c8b-8732-225a27ff1cec.png)

That's how it should look like:

![Listview (1)](https://user-images.githubusercontent.com/3354500/129206496-c99c40de-b6e3-414c-b0b9-e272ee6e4043.png)

Below the toplevel element, the pages are shown. In the current state, the text snippets are shown again. That's perfect.

beatrycze-volk replied:
Yes there is a problem with top element. It does't return any result if fulltext is empty as it throws error in SOLR Plugin.

So now basically plugin allows to index empty full text but return error when you search for document with such an empty full text field.

albig reviewed (changes requested):
Except for the mentioned issue in the listview, I stumbled about the word highlight.

I used the following GET-Parameter with your branch and with master. In your branch, no highlighting is active. Is this another bug or is this a missing configuration in my setup.

`tx_dlf[highlight_word]=Pieschen&tx_dlf[page]=22`

![Digitale Sammlungen  Auf der Suche nach Zukunft  Das Beispiel Pieschen (1)](https://user-images.githubusercontent.com/3354500/129206750-86ea269b-60c1-4521-beca-0aa44f9bfd4a.png)

![Digitale Sammlungen  Auf der Suche nach Zukunft  Das Beispiel Pieschen](https://user-images.githubusercontent.com/3354500/129206738-40217928-8afa-485c-8a0b-f65e3ddc9bc8.png)

beatrycze-volk commented:
> Except for the mentioned issue in the listview, I stumbled about the word highlight.
> 
> I used the following GET-Parameter with your branch and with master. In your branch, no highlighting is active. Is this another bug or is this a missing configuration in my setup.
> 
> `tx_dlf[highlight_word]=Pieschen&tx_dlf[page]=22`
> 
In my branch highlight happens by coordinates given in result document in element 'highlight'.

albig commented:
> > `tx_dlf[highlight_word]=Pieschen&tx_dlf[page]=22`
> 
> In my branch highlight happens by coordinates given in result document in element 'highlight'.

I don't understand how to make this work:

1. Search -->
2. Listview (search results) -->
3. PageView with word highlighting

beatrycze-volk commented:
> I don't understand how to make this work:
> 
> 1. Search -->
> 2. Listview (search results) -->
> 3. PageView with word highlighting

After you click one of the hits in Listview and you will get redirected to certain document, the method `triggerSearchAfterHitLoad()` should be fired. This method will check if URL contains `tx_dlf[highlight_word]` parameter. If it exists and has value, search for this value will be executed. The returned result document will contain coordinates for highlighting,

The method which is drectly responsible for displaying highlights in image is called `addImageHighlight(data)`. This one will check if there is anything to highlight and if yes it will notify viewer. Then viewer will highlight words matching to coordinates stored in `element['highlight'].`

albig commented:
> > I don't understand how to make this work:
> > 
> > 1. Search -->
> > 2. Listview (search results) -->
> > 3. PageView with word highlighting
> 
> After you click one of the hits in Listview and you will get redirected to certain document, the method `triggerSearchAfterHitLoad()` should be fired. This method will check if URL contains `tx_dlf[highlight_word]` parameter. If it exists and has value, search for this value will be executed. The returned result document will contain coordinates for highlighting,
> 
> The method which is drectly responsible for displaying highlights in image is called `addImageHighlight(data)`. This one will check if there is anything to highlight and if yes it will notify viewer. Then viewer will highlight words matching to coordinates stored in `element['highlight'].`

Thank you for your explanation.

So, the text highlighting only works with the SearchInDocument plugin used together with the pageview plugin AND the TypoScript configuration for the "Toolbox" must be included.

Nevertheless, it's still not working. I'll add some remarks in the following review on the right lines.

albig commented on Resources/Public/Javascript/PageView/SearchInDocument.js:
```diff
@@ -43,10 +43,226 @@ function resetStart() {
+    var page = getCurrentPage();
+
+    data['documents'].forEach(function (element, i) {
+        if(element['page'] === page) {
```
```suggestion
        if(element['page'] == page) {
```
In my Chromium console, the `page` is of type string. But the `element['page'] ` is of type integer. The comparison always fails.

Not sure, if it is the best way to change the comparison or fix the type of the `page` variable.

beatrycze-volk replied:
I would convert string to number and compare them.

beatrycze-volk commented:
> So, the text highlighting only works with the SearchInDocument plugin used together with the pageview plugin AND the TypoScript configuration for the "Toolbox" must be included.

Currently yes. There is also possibility to implement it in the way that coordinates will be already passed in URL parameter.


> Nevertheless, it's still not working. I'll add some remarks in the following review on the right lines.

Thanks!

albig commented on Resources/Public/Javascript/PageView/SearchInDocument.js:
```diff
@@ -43,10 +43,226 @@ function resetStart() {
+    for(var i = 0; i < queryParams.length; i++) {
+        var queryParam = queryParams[i].split('=');
+
+        if(queryParam[0] === $("input[id='tx-dlf-search-in-document-page']").attr('name')) {
```
This comparison doesn't respect the possible encoding of the parameters. This is wrong on other places, too.

If you click next/last page, the link parameter get's encoded:
<code>
/workview?tx_dlf%5Bdouble%5D=0&tx_dlf%5Bhighlight_word%5D=pieschen&tx_dlf%5Bid%5D=1&tx_dlf%5Bpage%5D=3&cHash=04b8a699974161fd87b83c9b36d65483
</code>

But `'tx_dlf%5Bpage%5D' === 'tx_dlf[page]'` is not true so the highlighting fails.

albig replied:
Highlighting seems to work now. Thank you!

albig reviewed (changes requested):
Please have look at my comments.

beatrycze-volk committed:
Apply review comments related to URL parameters

albig commented on Classes/Common/DocumentList.php:
```diff
@@ -316,6 +252,122 @@ protected function getRecord($element)
+                $fields['fulltext'] . ':' . Solr::escapeQuery($this->metadata['searchString']));
+                // TODO:toplevel documents throw error in SOLR PLugin:
+                // java.lang.RuntimeException: Could not determine OCR format for sample ''
+                //. ' OR ' . $fields['toplevel'] . ':true');
```
```suggestion
```
Just remove these lines. If you do a fulltext search, it is done already done by `$this->metadata['options']['select']` in line 292. Here you only overwrite this query again.

albig commented on Classes/Common/DocumentList.php:
```diff
@@ -316,6 +252,122 @@ protected function getRecord($element)
+        }
+        // Add filter query to get all documents with the required uid.
+        $filterQuery['key'] = $fields['uid'] . ':' . Solr::escapeQuery($record['uid']);
+        $query->addFilterQuery($filterQuery);
```
```suggestion
        $query->createFilterQuery('uid')->setQuery($fields['uid'] . ':' . Solr::escapeQuery($record['uid']));
```
This doesn't work and will be ignored. As result, you always search fulltext:(something) OR toplevel:true and receive ALL toplevel documents. But you only want the toplevel element for the given uid.

`addFilterQuery()` expects an array like this:
`$filterQuery = [ 'key' => 'fq1', 'query' => 'uid:1' ];`

So I'd suggest to use `createFilterQuery()` like in the [documentation of Solarium](https://solarium.readthedocs.io/en/4.2.0/queries/select-query/building-a-select-query/adding-filterqueries/). Maybe you find another solution.

albig commented on Classes/Common/DocumentList.php:
```diff
@@ -316,6 +252,122 @@ protected function getRecord($element)
+        foreach ($filterQueries as $key => $value) {
+            if (isset($value['query'])) {
+                $filterQuery[$key] = $value['query'] . ' OR ' . $fields['toplevel'] . ':true';
+                $query->addFilterQuery($filterQuery);
```
This won't work either. But I found no way to test this. I think, there is another issue in `Search:makeFacetsMenuArray()`. So I can't get the available facets from Solr and show the facets menu.

beatrycze-volk replied:
I have taken look to this method. Do you think that it needs to be adjusted now or it can be done later?

albig replied:
The addFilterQuery() above has to be fixed. But before doing this, I'm looking for working facets. Currently, the Solr has no facets available and I still don't know, why.

A query like 

`/solr/dlfCore0/select?facet=on&q=*%3A*&rows=0`

Should return something like

<pre>
{
  "responseHeader":{
    "status":0,
    "QTime":2,
    "params":{
      "q":"*:*",
      "rows":"0",
      "facet":"on"}},
  "response":{"numFound":4517,"start":0,"numFoundExact":true,"docs":[]
  },
  "facet_counts":{
    "facet_queries":{},
    "facet_fields":{},
    "facet_ranges":{},
    "facet_intervals":{},
    "facet_heatmaps":{}}}
</pre>

In my case - with this PR working - I get

<pre>
{
  "responseHeader":{
    "status":0,
    "QTime":0,
    "params":{
      "q":"*:*",
      "rows":"0",
      "facet":"on"}},
  "response":{"numFound":4376,"start":0,"numFoundExact":true,"docs":[]
  }}
</pre>

That's why no facetting is shown.

albig replied:
The faceting is solved (see other commit suggest).

With selecting a facet, we get the (expected) exception:

"A filterquery must have a key value" in Line 283.

The used GET-parameter are:

`tx_dlf[fq][0]=collection_faceting:("Projekt: Illustrierte Magazine der Klassischen Moderne")&tx_dlf[fq][1]=type_faceting:("volume")&tx_dlf[query]=*`

So please rewrite Line 283.

albig commented on Classes/Common/DocumentList.php:
```diff
@@ -316,6 +252,122 @@ protected function getRecord($element)
+        // Set additional query parameters.
+        // Set reasonable limit for safety reasons.
+        // We don't expect to get more than 10.000 hits per UID.
+        $query->setStart(0)->setRows(1000);
```
```suggestion
        $query->setStart(0)->setRows(10000);
```
Fix the limit.

albig reviewed (changes requested):
With the patched ocr-highlighting plugin, I see the toplevel element. But there are some issues with the facets and the filter queries. Please see my comments.

beatrycze-volk committed:
Remove overwrite of query

Co-authored-by: Alexander Bigga <alexander.bigga@slub-dresden.de>

beatrycze-volk committed:
Use create filter query

Co-authored-by: Alexander Bigga <alexander.bigga@slub-dresden.de>

beatrycze-volk committed:
Fix the query limit

Co-authored-by: Alexander Bigga <alexander.bigga@slub-dresden.de>

albig commented on Classes/Common/Indexer.php:
```diff
@@ -443,10 +448,10 @@ protected static function processLogical(Document &$doc, array $logicalUnit)
-        ) {
+        $logger = GeneralUtility::makeInstance('TYPO3\CMS\Core\Log\LogManager')->getLogger(__CLASS__);
+
+        $logger->error($doc->hasFulltext && $fullText = $doc->getFullText($physicalUnit['id']));
```
This line is not needed, isn't it?
```suggestion
```

Commit:
Update Classes/Common/Indexer.php

This line is not needed.

albig commented on Configuration/ApacheSolr/configsets/dlf/conf/solrconfig.xml:
```diff
@@ -700,6 +708,12 @@
+
+    <arr name="components">
+          <str>query</str>
+          <str>ocrHighlight</str>
```
```suggestion
          <str>facet</str>
          <str>ocrHighlight</str>
```
We use the facetComponent for search facetting.

albig commented on Classes/Plugin/Eid/SearchInDocument.php:
```diff
@@ -58,21 +59,37 @@ public function main(ServerRequestInterface $request)
+
+            $response = $solr->service->executeRequest($solrRequest);
+            $result = $solr->service->createResult($query, $response);
+            $output['numFound'] = $result->getNumFound();
```
```suggestion
            /** @scrutinizer ignore-call */
            $output['numFound'] = $result->getNumFound();
```
This is a false-positiv and should be ignored.

albig reviewed (changes requested):
Maybe the last change request for now. Please have a look at the comments.

beatrycze-volk committed:
Ignore false-positive warning from Scrutinizer

Co-authored-by: Alexander Bigga <alexander.bigga@slub-dresden.de>

beatrycze-volk committed:
Add facet component to solrconfig.xml

Co-authored-by: Alexander Bigga <alexander.bigga@slub-dresden.de>

beatrycze-volk commented:
> With selecting a facet, we get the (expected) exception:
> 
> "A filterquery must have a key value" in Line 283.

Is `$key` not a key value?

albig commented on Classes/Common/DocumentList.php:
```diff
@@ -316,6 +252,116 @@ protected function getRecord($element)
+        // Extend filter query to get all documents with the same UID.
+        foreach ($filterQueries as $key => $value) {
+            if (isset($value['query'])) {
+                $filterQuery[$key] = $value['query'] . ' OR ' . $fields['toplevel'] . ':true';
```
```suggestion
                $filterQuery[$key] = $value['query'] . ' OR ' . $fields['toplevel'] . ':true';
                $filterQuery = [
                    'key' => $key,
                    'query' => $value['query'] . ' OR ' . $fields['toplevel'] . ':true'
                ];
```

Solarium expects an array like this at this point. Important is the key "key". The key "query" seems not to matter.

The other approach would be in the above foreach loop:

`$query->createFilterQuery('kitodo' . $key)->setQuery($value['query'] . ' OR ' . $fields['toplevel'] . ':true');`

This calls addFilterQuery internally.

beatrycze-volk replied:
Oh, now I understand. Thanks!

beatrycze-volk committed:
Use array for filter queyr

Co-authored-by: Alexander Bigga <alexander.bigga@slub-dresden.de>

Commit:
Merge branch 'master' into use-solr-highlighting

albig commented:
I think we finally reached the state, all features seems to work and I can have a look at the performance of the new ocr highlighting plugin and about the "migration" path from last release to next release containing this feature. So maybe, there is coming another change request - or the final merge in new future.

albig commented:
Just to give an update.

Indexing with the same set of documents seems to work with this PR. But there are missing some pages. Obviously the ocrhighlight plugin doesn't like pages with the `<Illustration>` tag.

Example:
https://digital.slub-dresden.de/data/kitodo/lebe_364572701-19320800/lebe_364572701-19320800_ocr/00000068.xml
<pre>
at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.lang.RuntimeException: Failed to parse the OCR markup, make sure your files are well-formed and your regions start/end on complete tags! (Source was: [unknown])
	at de.digitalcollections.solrocr.formats.OcrParser.<init>(OcrParser.java:109)
	at de.digitalcollections.solrocr.formats.alto.AltoParser.<init>(AltoParser.java:27)
	at de.digitalcollections.solrocr.formats.alto.AltoFormat.getParser(AltoFormat.java:38)
	at de.digitalcollections.solrocr.model.OcrFormat.filter(OcrFormat.java:90)
	at de.digitalcollections.solrocr.lucene.filters.OcrCharFilterFactory.create(OcrCharFilterFactory.java:51)
</pre>

The other point is, that the index size is 3 times (!) bigger than before.

albig commented:
> Just to give an update.

I'm again on this PR. Just a new update.

> Indexing with the same set of documents seems to work with this PR. But there are missing some pages. Obviously the ocrhighlight plugin doesn't like pages with the `<Illustration>` tag.

It has been to very long `<Polygon>` element. See https://github.com/dbmdz/solr-ocrhighlighting/issues/210. This has been fixed.

> The other point is, that the index size is 3 times (!) bigger than before.

Storing the whole ALTO is not that nice. This PR ignores additionally the ALTO namespace. Not sure, if we should go this way.

If we store only the suggested miniOcr-format, the index has "only" double size than before.

beatrycze-volk committed:
Remove Illustration tag from full text XML

beatrycze-volk commented:
> > Just to give an update.
> 
> I'm again on this PR. Just a new update.
> 
> > Indexing with the same set of documents seems to work with this PR. But there are missing some pages. Obviously the ocrhighlight plugin doesn't like pages with the `<Illustration>` tag.
> 
> It has been to very long `<Polygon>` element. See [dbmdz/solr-ocrhighlighting#210](https://github.com/dbmdz/solr-ocrhighlighting/issues/210). This has been fixed.

I was actually also on this PR. Right after push my 'fix' commit I have seen this comment. Probably I will need to delete it.

> > The other point is, that the index size is 3 times (!) bigger than before.
> 
> Storing the whole ALTO is not that nice. This PR ignores additionally the ALTO namespace. Not sure, if we should go this way.
> 
> If we store only the suggested miniOcr-format, the index has "only" double size than before.

Ok. Would it be our approach now to convert files to miniOCR? Should it be done in our code or by some external tool?

Commit:
Merge branch 'master' into use-solr-highlighting

albig commented:
> I was actually also on this PR. Right after push my 'fix' commit I have seen this comment. Probably I will need to delete it.

Yes, I think so. This should'nt be fixed on our side.

> Ok. Would it be our approach now to convert files to miniOCR? Should it be done in our code or by some external tool?

I've done this in our code because this miniOCR is no standard format.

I'd suggest the following procedure to continue:

1. I merge #587 and #639 into a new feature branch `add-ocrhighlighting-plugin`
2. I'll add my prototype code for miniOCR to this branch
3. We can discuss the best solution in a new PR `add-ocrhighlighting-plugin` -> `master`

beatrycze-volk commented:
> 1. I merge [Use Solr OCR Highlighting Plugin in Search in Document Plugin #587](https://github.com/kitodo/kitodo-presentation/pull/587) and [Remove obsolete hl parameter use in image highlighting #639](https://github.com/kitodo/kitodo-presentation/pull/639) into a new feature branch `add-ocrhighlighting-plugin`
> 2. I'll add my prototype code for miniOCR to this branch
> 3. We can discuss the best solution in a new PR `add-ocrhighlighting-plugin` -> `master`

Ok!
